### PR TITLE
Activities integration

### PR DESCRIPTION
**Pull request by oliveiraigorm**

### CODE REVIEW

**Comment by oliveiraigorm** [comment]
@mwalbeck Did a lot of debugging and looking at examples but couldn't get the activity posted, could you latter take a look on what might be missing? Thanks!

**Comment by mwalbeck** [comment]
Okay, so I have taken a look, and from what I can see I think you're missing a Provider. So there are 3 components, the filter, the provider and the settings. The provider is for displaying the actual activity. I can at least see in the database that the activity is actually being created. Here is an example from the dav app, which is the calendar backend https://github.com/nextcloud/server/blob/master/apps/dav/lib/CalDAV/Activity/Provider/Calendar.php

The provider also needs to be registered in the info.xml.

It seems the common approach is to use a more computer friendly string for the subject of the event and then use that to figure out the type of event and then in the provider put together a human-readable string to present in the activity app.

File: lib/Controller/PageController.php [hunk 1]
```diff
@@ -69,14 +74,32 @@ public function index(): TemplateResponse {
...
+
 			$this->mapper->insert($client);
+
+			$event = $this->activityManager->generateEvent();
+			$event->setApp('adminly_clients')
+				->setObject('client', $client->getId())
+				->setType('clients')
+				->setAffectedUser($this->userId)
+				->setSubject(
+					"New Client added",
```
Comment (mwalbeck) [inline lib/Controller/PageController.php:92]: The subject could be changed to something like client_added and then in the provider file that is turned into something more human-readable to present to the user.

**Comment by oliveiraigorm** [comment]
> Okay, so I have taken a look, and from what I can see I think you're missing a Provider. So there are 3 components, the filter, the provider and the settings. The provider is for displaying the actual activity. I can at least see in the database that the activity is actually being created. Here is an example from the dav app, which is the calendar backend https://github.com/nextcloud/server/blob/master/apps/dav/lib/CalDAV/Activity/Provider/Calendar.php
> 
> The provider also needs to be registered in the info.xml.
> 
> It seems the common approach is to use a more computer friendly string for the subject of the event and then use that to figure out the type of event and then in the provider put together a human-readable string to present in the activity app.

Thanks! I was thinking the Provider was necessary for more complex rendering and not mandatory, my mistake.

File: lib/Controller/PageController.php (hunk 1 above)
Comment (oliveiraigorm) [reply lib/Controller/PageController.php]: Managed to get it almost there, I'm able to display it but need to do some fixes to make the client name bold.

image

**Comment by oliveiraigorm** [comment]
@mwalbeck 

Currently I'm able to show the activities like this:

image

To the the client name shown in bold there's a need to set the RichSubject of the event by sending the parameter(s) (the client object in this case). I've tried to set it in multiple ways, following the files and comments activities, but with no success.  

image

Doing a lot of debugging I found out that this error is thrown. I've tried to add a client entry to Definitions.php but that didn't solve it either. 

image

**Comment by mwalbeck** [comment]
@oliveiraigorm Very nice job so far, I'll take a look and see if I can make it behave

File: lib/Controller/PageController.php [hunk 2]
```diff
@@ -69,14 +74,33 @@ public function index(): TemplateResponse {
...
+							'name' => $client->getName()
+						],
+					]
+				);
+
+			$this->activityManager->publish($event);
+
 			return "Success";
 		} catch (Exception $e) {
 			return $e->getMessage();
```
Comment (mwalbeck) [inline lib/Controller/PageController.php:106]: This should probably be changed to `throw $e;`

File: lib/Controller/PageController.php [hunk 3]
```diff
@@ -69,14 +74,33 @@ public function index(): TemplateResponse {
...
-		$client->setProviderId($this->userId);
-
 		try {
+			$client = new Client();
+			$client->setName($name);
+			$client->setEmail($email);
+			$client->setDescription($description);
+			$client->setProviderId($this->userId);
+
 			$this->mapper->insert($client);
```
Comment (mwalbeck) [inline lib/Controller/PageController.php:84]: We could store the return value in a variable called $new_client and then return that at the end for a more useful return value. That will also follow conventions more.

File: lib/Controller/PageController.php [hunk 4]
```diff
@@ -69,14 +74,33 @@ public function index(): TemplateResponse {
...
+							'type' => 'addressbook-contact',
+							'id' => $client->getId(),
+							'name' => $client->getName()
+						],
+					]
+				);
+
+			$this->activityManager->publish($event);
+
 			return "Success";
```
Comment (mwalbeck) [inline lib/Controller/PageController.php:104]: return $new_client;

File: lib/Activity/Provider.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	/**
+	 * @param string $language
+	 * @param IEvent $event
+	 * @param IEvent|null $previousEvent
+	 * @return IEvent
+	 * @throws \InvalidArgumentException
+	 * @since 11.0.0
+	 */
+	public function parse($language, IEvent $event, IEvent $previousEvent = null) {
+		$subjectParameters = $event->getSubjectParameters();
```
Comment (mwalbeck) [inline lib/Activity/Provider.php:80]: We should move this one below the if statement, so we don't actually store the parameters unless it's an event we're going to process.

File: lib/Activity/Provider.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	/** @var IManager */
+	protected $activityManager;
+
+	/** @var IUrlGenerator */
+
+	protected $url;
+
+	/** @var IEventMerger */
+
+	protected $eventMerger;
```
Comment (mwalbeck) [inline lib/Activity/Provider.php:55]: For now the eventMerger isn't being used, so let's remove it.

File: lib/Activity/Provider.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+	/**
+	 * @param IFactory $languageFactory
+	 * @param IURLGenerator $url
+	 * @param IManager $activityManager
+	 * @param IGroupManager $groupManager
```
Comment (mwalbeck) [inline lib/Activity/Provider.php:61]: GroupManager also isn't being used so let's remove the comment.

**Review by mwalbeck — CHANGES_REQUESTED** [review]

File: lib/Controller/PageController.php (hunk 3 above)
Comment (mwalbeck) [reply lib/Controller/PageController.php:84]: Oh, I meant something slightly different. If change all the $new_client variables back to $client and then just change line 84 to

```
$new_client = $this->mapper->insert($client);
```

And then keep the `return $new_client` on line 104

**Review by mwalbeck — APPROVED** [review]

**Pull request merged**